design/storage-class-device-set.md

```diff
@@ -0,0 +1,487 @@
+storage that can be dynamically provisioned and attached to any node in a given
+Availability Zone (AZ). The design can also accommodate non-cloud environments
+through the use of local PVs.
+
+## StorageClassDeviceSet struct
+
+```go
+// File: pkg/apis/ceph.rook.io/v1/types.go
+package v1
+
+import (
+    v1 "k8s.io/api/core/v1"
+
+    rook "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+)
+
+struct StorageClassDeviceSet {
+    Name                 string // A unique identifier for the set
+    Count                int // Number of devices in this set
+    Resources            v1.ResourceRequirements // Requests/limits for the devices
+    Placement            rook.Placement // Placement constraints for the devices
+    Config               map[string]string // Provider-specific device configuration
+    volumeClaimTemplates []v1.PersistentVolumeClaim // List of PVC templates for the underlying storage devices
+}
+```
+
+A provider will be able to use the `StorageClassDeviceSet` struct to describe
+the properties of a particular set of `StorageClassDevices`. In this design, the
+notion of a "`StorageClassDevice`" is an abstract concept, separate from
+underlying storage devices. There are three main aspects to this abstraction:
+
+1. A `StorageClassDevice` is both storage and the software required to make the
+storage available and manage it in the cluster. As such, the struct takes into
+account the resources required to run the associated software, if any.
+1. A single `StorageClassDevice` could be comprised of multiple underlying
+storage devices, specified by having more than one item in the
+`volumeClaimTemplates` field.
+1. Since any storage devices that are part of a `StorageClassDevice` will be
+represented by block-mode PVCs, they will need to be associated with a Pod so
+that they can be attached to cluster nodes.
+
+A `StorageClassDeviceSet` will have the following fields associated with it:
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no
+  resource requests.
+* **placement**: The placement criteria for the devices. Default is no
+  placement criteria.
+* **config**: Granular device configuration. This is a generic
+  `map[string]string` to allow for provider-specific configuration.
+* **volumeClaimTemplates**: A list of PVC templates to use for provisioning the
+  underlying storage devices.
+
+An entry in `volumeClaimTemplates` must specify the following fields:
+  * **resources.requests.storage**: The desired capacity for the underlying
+  storage devices.
+  * **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
+
+The CephCluster CRD could be extended to include a new field:
+`spec.storage.StorageClassDeviceSets`, which would be a list of one or more
+`StorageClassDeviceSets`. If elements exist in this list, the CephCluster
```

Comment:
This should be `StorageClassDeviceSet` (singular), right?

Reply:
"one or more `StorageClassDeviceSet`" is grammatically awkward, and "`StorageClassDeviceSet`s" just looks weird.

Reply:
how about "a list of one or more items of type `StorageClassDeviceSet`"

Reply:
I think it'll be fine. I mean, heck, the title of this doc is `# Rook StorageClassDeviceSets`. ;)